integration-tests/validation/src/main/java/com/android/designcompose/testapp/validation/examples/BrushTest.kt

```diff
@@ -77,40 +93,82 @@ class SizingShaderBrush(private val shader: RuntimeShader) : ShaderBrush() {
-                    animation = tween(10 * 1000, easing = FastOutSlowInEasing),
-                    repeatMode = RepeatMode.Reverse
-                )
+    val shaderEnabled = remember { mutableStateOf(false) }
+    val solidColor = SolidColor(Color.Blue)
+    val linearGradient = Brush.linearGradient(listOf(Color.Red, Color.Yellow, Color.Green))
+    val radialGradient = Brush.radialGradient(listOf(Color.Red, Color.Yellow, Color.Green))
+    val brushState: MutableState<Brush> = remember { mutableStateOf(solidColor) }
+    CustomBrushTestDoc.Main(fill = brushState)
+
+    Row(modifier = Modifier.offset(y = 300.dp)) {
+        ElevatedTextButton(
+            onClick = {
+                shaderEnabled.value = false
+                brushState.value = solidColor
+            },
+            text = "Solid Blue"
+        )
+        ElevatedTextButton(
+            onClick = {
+                shaderEnabled.value = false
+                brushState.value = linearGradient
+            },
+            text = "Linear"
+        )
+        ElevatedTextButton(
+            onClick = {
+                shaderEnabled.value = false
+                brushState.value = radialGradient
+            },
+            text = "Radial"
         )
 
-    // Android T introduces AGSL and RuntimeShader. Robolectric (part of our test infrastructure)
-    // only supports software rendering.
-    val brush: () -> Brush =
+        // Android T introduces AGSL and RuntimeShader. Robolectric (part of our test
+        // infrastructure) only supports software rendering.
         if (
             Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
                 "robolectric" != Build.FINGERPRINT
         ) {
+            val infiniteTransition = rememberInfiniteTransition(label = "animate shader")
+            val movingValue =
+                infiniteTransition.animateFloat(
+                    label = "moving value for shader",
+                    initialValue = 0.0f,
+                    targetValue = 10.0f,
+                    animationSpec =
+                        infiniteRepeatable(
+                            animation = tween(10 * 1000, easing = FastOutSlowInEasing),
+                            repeatMode = RepeatMode.Reverse
+                        )
+                )
             val shader = RuntimeShader(CUSTOM_SHADER)
-            val shaderBrush = SizingShaderBrush(shader)
-            // The kotlin compiler seems to get confused without semicolons on these statements.
-            shader.setFloatUniform("iTime", 0.0f);
-
-            {
-                // Only sample the state in the generator function; this means that we avoid
-                // recomposition and only do the redraw phase.
-                shader.setFloatUniform("iTime", movingValue.value)
-                shaderBrush
+            shader.setFloatUniform("iTime", 0.0f)
+            LaunchedEffect(shaderEnabled.value) {
+                if (shaderEnabled.value) {
```

Comment:
I like this better than what you have before, but it is still a little confusing because I am not familiar with `snapshotFlow`. I looked it up and it makes more sense. Can you add a comment here explaining what is going on? Other than that I am good with this, though @iamralpht should take another look.

Reply:
Done. Added a comment.
@iamralpht Please take another look and let me know what you think, thanks!

Reply:
I'm still a bit confused about what is getting re-run when `movingValue.value` changes. In the previous version, it looks like it mutated `shaderBrush` (by giving it a new value for the `iTime` uniform) and returned it, and I think just that generator function would get re-run when `movingValue.value` changed (and then just Compose's "redraw" phase would run, skipping Composition and Layout).

Here, I think that there might be a bit more stuff that's getting run... but I'm not sure. I am guessing that you have to make a new instance of `SizingShaderBrush` because otherwise just mutating the `iTime` uniform doesn't change the identity of the instance already in `brushState.value` (so probably the `State<>` object just ignores assigning the value to the current value and doesn't propagate an invalidation).

So the usage of this seems more involved and complicated than before, and I think it might be less efficient than before, too. Right now, I would say we shouldn't make this change because the previous version was good in terms of efficiency (where the previous versions of the other things that didn't take `State<>` was quite inefficient and the move to `State<>` was a performance win).

Reply:
Right now when `moving.value changes` it has to create a new `shaderBrush` to trigger recomposition. So I agree it might be less efficient than before. I will close this pr. Thanks very much for the feedback!